ethcore/src/trace/db.rs

```diff
@@ -0,0 +1,573 @@
+				.map(|block_traces| block_traces.bloom())
+				.map(BlockTracesBloom::from)
+				.map(Into::into)
+				.collect();
+
+			let chain = BloomGroupChain::new(self.config.blooms, self);
+			let trace_blooms = chain.replace(&replaced_range, enacted_blooms);
+			let blooms_to_insert = trace_blooms.into_iter()
+				.map(|p| (From::from(p.0), From::from(p.1)))
+				.collect::<HashMap<TraceGroupPosition, BlockTracesBloomGroup>>();
+
+			let mut blooms = self.blooms.write().unwrap();
+			batch.extend_with_cache(&mut blooms, blooms_to_insert, CacheUpdatePolicy::Remove);
+		}
+
+		self.tracesdb.write(batch).unwrap();
+	}
+
+	fn trace(&self, block_number: BlockNumber, tx_position: usize, trace_position: usize) -> Option<LocalizedTrace> {
+		self.extras.block_hash(block_number)
+			.and_then(|block_hash| self.traces(&block_hash)
+				.map(FlatBlockTraces::from)
+				.map(Into::<Vec<FlatTransactionTraces>>::into)
+				.and_then(|traces| traces.into_iter().nth(tx_position))
+				.map(Into::<Vec<FlatTrace>>::into)
+				.and_then(|traces| traces.into_iter().nth(trace_position))
+				.map(|trace| {
+					let tx_hash = self.extras.transaction_hash(block_number, tx_position)
+						.expect("Expected to find transaction hash. Database is probably malformed");
+
+					LocalizedTrace {
+						parent: trace.parent,
+						children: trace.children,
+						depth: trace.depth,
+						action: trace.action,
+						result: trace.result,
+						trace_number: trace_position,
+						transaction_number: tx_position,
+						transaction_hash: tx_hash,
+						block_number: block_number,
+						block_hash: block_hash,
+					}
+				})
+			)
+	}
+
+	fn transaction_traces(&self, block_number: BlockNumber, tx_position: usize) -> Option<Vec<LocalizedTrace>> {
+		self.extras.block_hash(block_number)
+			.and_then(|block_hash| self.traces(&block_hash)
+				.map(FlatBlockTraces::from)
+				.map(Into::<Vec<FlatTransactionTraces>>::into)
+				.and_then(|traces| traces.into_iter().nth(tx_position))
+				.map(Into::<Vec<FlatTrace>>::into)
+				.map(|traces| {
+					let tx_hash = self.extras.transaction_hash(block_number, tx_position)
+						.expect("Expected to find transaction hash. Database is probably malformed");
+
+					traces.into_iter()
+					.enumerate()
+					.map(|(i, trace)| LocalizedTrace {
+						parent: trace.parent,
+						children: trace.children,
+						depth: trace.depth,
+						action: trace.action,
+						result: trace.result,
+						trace_number: i,
+						transaction_number: tx_position,
+						transaction_hash: tx_hash.clone(),
+						block_number: block_number,
+						block_hash: block_hash
+					})
+					.collect()
+				})
+			)
+	}
+
+	fn block_traces(&self, block_number: BlockNumber) -> Option<Vec<LocalizedTrace>> {
+		self.extras.block_hash(block_number)
+			.and_then(|block_hash| self.traces(&block_hash)
+				.map(FlatBlockTraces::from)
+				.map(Into::<Vec<FlatTransactionTraces>>::into)
```

Comment:
3 lines above are quite common (6 occurrences) (`traces.map(FlatBlock).map(Into::Vec<...`) maybe it should be separated into method?